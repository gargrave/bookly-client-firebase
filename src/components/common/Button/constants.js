export const POSITIONS = [
  'left',
  'right',
];

export const TYPES = [
  'success',
  'secondary',
  'info',
  'warning',
  'danger',
  'light',
  'dark',
];
