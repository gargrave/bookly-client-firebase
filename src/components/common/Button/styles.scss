.bookly-button {
  height: 2.15rem;
  line-height: 2rem;
  margin-left: 10px;
  padding: 0 8px;

  &--left {
    margin-left: 0;
  }

  &--right {
    margin-left: auto;
  }
}
