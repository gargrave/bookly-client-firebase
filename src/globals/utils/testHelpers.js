import React from 'react';
import { shallow } from 'enzyme';

export class ComponentBuilder {
  constructor(component, defaultProps) {
    this.component = component;
    this.defaultProps = defaultProps;
  }

  shallowGetComponent(extraProps = {}) {
    const props = {
      ...this.defaultProps,
      ...extraProps,
    };
    return shallow(<this.component {...props} />);
  }
}
