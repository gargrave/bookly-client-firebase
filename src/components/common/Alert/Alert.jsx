// @flow
import React from 'react';
import { any, oneOf, string } from 'prop-types';

import { buildClasses } from '../../../globals/utils/cssHelpers';

type Props = {
  children?: any,
  message: string,
  type: string,
};

const Alert = ({
  children,
  message,
  type,
}: Props) => (
  <div className={buildClasses('alert', ['alert', `alert-${type}`])}>
    {message}
    {children}
  </div>
);

Alert.propTypes = {
  children: any,
  message: string.isRequired,
  type: oneOf([
    'primary',
    'secondary',
    'success',
    'info',
    'warning',
    'danger',
    'light',
    'dark',
  ]).isRequired,
};

export default Alert;
