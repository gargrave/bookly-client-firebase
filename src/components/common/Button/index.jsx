// @flow
import React from 'react';
import { bool, func, oneOf, string } from 'prop-types';

import { buildClasses } from '../../../utils/cssHelpers';

import './styles.css';

type Props = {
  canSubmit?: boolean,
  classes?: string,
  disabled?: boolean,
  onClick: Function,
  position?: string,
  text: string,
  type?: string,
};

const acceptableTypes = [
  'success',
  'secondary',
  'info',
  'warning',
  'danger',
  'light',
  'dark',
];

const acceptablePositions = [
  'left',
  'right',
];

const buttonClass = (type: string) => {
  if (acceptableTypes.includes(type)) {
    return `button-${type}`;
  }
  return '';
};

const positionClass = (position: string) => {
  if (acceptablePositions.includes(position)) {
    return `button--${position}`;
  }
  return '';
};

const buildClassList = (
  type: string,
  position: string,
  classes: string,
) => {
  return buildClasses(
    ['button', positionClass(position)],
    [buttonClass(type), ...classes.split(' ')],
  );
};

const Button = ({
  canSubmit,
  classes = '',
  disabled,
  onClick,
  position = '',
  text,
  type,
}: Props) => {
  return (
    <button
      className={buildClassList(type || '', position, classes)}
      disabled={disabled || false}
      onClick={onClick}
      type={canSubmit ? 'submit' : 'button'}
    >
      {text}
    </button>
  );
};

Button.propTypes = {
  canSubmit: bool,
  classes: string,
  disabled: bool,
  onClick: func.isRequired,
  position: oneOf(acceptablePositions),
  text: string.isRequired,
  type: oneOf(acceptableTypes),
};

export default Button;
