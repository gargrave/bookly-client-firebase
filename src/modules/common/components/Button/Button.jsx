// @flow
import React from 'react'
import { bool, func, oneOf, string } from 'prop-types'

import { buildClass } from '../../../../utils/cssHelpers'

import styles from './Button.css'

export const POSITIONS = ['left', 'right']

export const TYPES = [
  'success',
  'secondary',
  'info',
  'warning',
  'danger',
  'light',
  'dark',
]

type Props = {
  canSubmit?: boolean,
  disabled?: boolean,
  onClick: Function,
  position?: string,
  text: string,
  type?: string,
}

const Button = ({
  canSubmit,
  disabled,
  onClick,
  position = '',
  text,
  type = '',
}: Props) => (
  <button
    className={buildClass(
      styles.button,
      { [styles[position]]: !!position },
      'button',
      { [`button-${type}`]: !!type },
    )}
    disabled={disabled || false}
    onClick={onClick}
    type={canSubmit ? 'submit' : 'button'}
  >
    {text}
  </button>
)

Button.propTypes = {
  canSubmit: bool,
  disabled: bool,
  onClick: func.isRequired,
  position: oneOf(POSITIONS),
  text: string.isRequired,
  type: oneOf(TYPES),
}

export default Button
