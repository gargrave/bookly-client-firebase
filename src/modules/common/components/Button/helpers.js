// @flow
import { buildClasses } from '../../../../globals/utils/cssHelpers';

import {
  POSITIONS,
  TYPES,
} from './constants';

const buttonClass = (type: string) => {
  if (TYPES.includes(type)) {
    return `button-${type}`;
  }
  return '';
};

const positionClass = (position: string) => {
  if (POSITIONS.includes(position)) {
    return `button--${position}`;
  }
  return '';
};

export const buildClassList = (
  type: string,
  position: string,
  classes: string,
) => {
  return buildClasses(
    ['button', positionClass(position)],
    [buttonClass(type), ...classes.split(' ')],
  );
};
