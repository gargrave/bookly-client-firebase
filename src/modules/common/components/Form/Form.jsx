// @flow
import React from 'react'
import { any, array, func, bool, string } from 'prop-types'

import { buildClass } from '../../../../utils/cssHelpers'

import Alert from '../Alert/Alert'
import Button from '../Button'
import ButtonRow from '../ButtonRow/ButtonRow'

import styles from './Form.css'

type Props = {
  cancelBtnText?: string,
  children?: any,
  classes?: string[],
  disabled?: boolean,
  onCancel?: Function,
  onSubmit: Function,
  submitBtnText?: string,
  submitDisabled?: boolean,
  topLevelError?: string,
}

const Form = ({
  cancelBtnText,
  children,
  classes = [],
  disabled,
  onCancel,
  onSubmit,
  submitBtnText,
  submitDisabled = false,
  topLevelError,
}: Props) => {
  return (
    <div>
      {topLevelError && <Alert message={topLevelError} type="danger" />}
      <form
        className={buildClass(styles.form, ...(classes || []))}
        onSubmit={onSubmit}
        noValidate
      >
        {children}

        <ButtonRow>
          <Button
            canSubmit={true}
            disabled={submitDisabled || disabled || false}
            onClick={onSubmit}
            position="left"
            text={submitBtnText || 'Submit'}
            type="success"
          />

          {onCancel && (
            <Button
              onClick={onCancel}
              position="right"
              text={cancelBtnText || 'Cancel'}
              type="light"
            />
          )}
        </ButtonRow>
      </form>
    </div>
  )
}

Form.propTypes = {
  cancelBtnText: string,
  children: any,
  classes: array,
  disabled: bool,
  onCancel: func,
  onSubmit: func.isRequired,
  submitBtnText: string,
  submitDisabled: bool,
  topLevelError: string,
}

export default Form
