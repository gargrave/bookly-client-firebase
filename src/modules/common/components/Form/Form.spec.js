import { ComponentBuilder } from '../../../../utils/testHelpers'

import Alert from '../../../common/components/Alert/Alert'
import Button from '../../../common/components/Button/Button'

import Form from './Form'

const defaultProps = {
  cancelBtnText: 'Cancel',
  children: [],
  classes: [],
  disabled: false,
  onCancel: jest.fn(),
  onSubmit: jest.fn(),
  submitBtnText: 'Submit',
  submitDisabled: false,
  topLevelError: '',
}

const builder = new ComponentBuilder(Form, defaultProps)

describe('Form', () => {
  let component

  it('matches the snapshot', () => {
    component = builder.shallowGetComponent()
    expect(component).toMatchSnapshot()
  })

  describe('basic rendering', () => {
    it('renders correctly', () => {
      component = builder.mountComponent()
      expect(component.find(Alert)).toHaveLength(0)
      expect(component.find(Button)).toHaveLength(2)
    })

    it('does not render "cancel" button if prop is empty', () => {
      component = builder.mountComponent({
        onCancel: null,
      })
      expect(component.find(Button)).toHaveLength(1)
    })

    it('renders an Alert if the prop is present', () => {
      component = builder.mountComponent({
        topLevelError: 'OMFG',
      })
      expect(component.find(Alert)).toHaveLength(1)
    })
  })
})
