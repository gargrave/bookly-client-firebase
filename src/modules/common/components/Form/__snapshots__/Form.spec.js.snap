// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form matches the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Form
    cancelBtnText="Cancel"
    classes={Array []}
    disabled={false}
    onCancel={[Function]}
    onSubmit={[Function]}
    submitBtnText="Submit"
    submitDisabled={false}
    topLevelError=""
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        "",
        <form
          className=""
          noValidate={true}
          onSubmit={[Function]}
        >
          <ButtonRow>
            <Button
              canSubmit={true}
              disabled={false}
              onClick={[Function]}
              position="left"
              text="Submit"
              type="success"
            />
            <Button
              onClick={[Function]}
              position="right"
              text="Cancel"
              type="light"
            />
          </ButtonRow>
        </form>,
      ],
    },
    "ref": null,
    "rendered": Array [
      "",
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            Array [],
            <ButtonRow>
              <Button
                canSubmit={true}
                disabled={false}
                onClick={[Function]}
                position="left"
                text="Submit"
                type="success"
              />
              <Button
                onClick={[Function]}
                position="right"
                text="Cancel"
                type="light"
              />
            </ButtonRow>,
          ],
          "className": "",
          "noValidate": true,
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": Array [
                <Button
                  canSubmit={true}
                  disabled={false}
                  onClick={[Function]}
                  position="left"
                  text="Submit"
                  type="success"
                />,
                <Button
                  onClick={[Function]}
                  position="right"
                  text="Cancel"
                  type="light"
                />,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "canSubmit": true,
                  "disabled": false,
                  "onClick": [Function],
                  "position": "left",
                  "text": "Submit",
                  "type": "success",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "onClick": [Function],
                  "position": "right",
                  "text": "Cancel",
                  "type": "light",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": [Function],
          },
        ],
        "type": "form",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          "",
          <form
            className=""
            noValidate={true}
            onSubmit={[Function]}
          >
            <ButtonRow>
              <Button
                canSubmit={true}
                disabled={false}
                onClick={[Function]}
                position="left"
                text="Submit"
                type="success"
              />
              <Button
                onClick={[Function]}
                position="right"
                text="Cancel"
                type="light"
              />
            </ButtonRow>
          </form>,
        ],
      },
      "ref": null,
      "rendered": Array [
        "",
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              Array [],
              <ButtonRow>
                <Button
                  canSubmit={true}
                  disabled={false}
                  onClick={[Function]}
                  position="left"
                  text="Submit"
                  type="success"
                />
                <Button
                  onClick={[Function]}
                  position="right"
                  text="Cancel"
                  type="light"
                />
              </ButtonRow>,
            ],
            "className": "",
            "noValidate": true,
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": Array [
                  <Button
                    canSubmit={true}
                    disabled={false}
                    onClick={[Function]}
                    position="left"
                    text="Submit"
                    type="success"
                  />,
                  <Button
                    onClick={[Function]}
                    position="right"
                    text="Cancel"
                    type="light"
                  />,
                ],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "canSubmit": true,
                    "disabled": false,
                    "onClick": [Function],
                    "position": "left",
                    "text": "Submit",
                    "type": "success",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "onClick": [Function],
                    "position": "right",
                    "text": "Cancel",
                    "type": "light",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
              ],
              "type": [Function],
            },
          ],
          "type": "form",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
